Mathlib/Data/Rat/Star.lean

```diff
@@ -1,52 +1,42 @@
 -/
+import Mathlib.Algebra.GroupWithZero.Power
 import Mathlib.Algebra.Star.Order
-import Mathlib.Data.Rat.Lemmas
-import Mathlib.Data.Rat.Order
-import Mathlib.GroupTheory.Submonoid.Membership
+import Mathlib.Algebra.Parity
 
 #align_import data.rat.star from "leanprover-community/mathlib"@"31c24aa72e7b3e5ed97a8412470e904f82b81004"
 
-/-! # Star order structure on ℚ
+/-!
+# Star ordered ring structure on ℚ
 
-Here we put the trivial `star` operation on `ℚ` for convenience and show that it is a
-`StarOrderedRing`. In particular, this means that every element of `ℚ` is a sum of squares.
+This file shows that `ℚ` is a `StarOrderedRing`. In particular, this means that every nonnegative
+rational number is a sum of squares.
 -/
 
+open AddSubmonoid Set
 
 namespace Rat
 
-instance : StarRing ℚ where
-  star := id
-  star_involutive _ := rfl
-  star_mul _ _ := mul_comm _ _
-  star_add _ _ := rfl
-
-instance : TrivialStar ℚ where star_trivial _ := rfl
-
-instance : StarOrderedRing ℚ :=
-  StarOrderedRing.ofNonnegIff (fun {_ _} => add_le_add_left) fun x => by
-    refine'
-      ⟨fun hx => _, fun hx =>
-        AddSubmonoid.closure_induction hx (by rintro - ⟨s, rfl⟩; exact mul_self_nonneg s) le_rfl
-          fun _ _ => add_nonneg⟩
-    /- If `x = p / q`, then, since `0 ≤ x`, we have `p q : ℕ`, and `p / q` is the sum of `p * q`
-      copies of `(1 / q) ^ 2`, and so `x` lies in the `AddSubmonoid` generated by square elements.
-
-      Note: it's possible to rephrase this argument as `x = (p * q) • (1 / q) ^ 2`, but this would
-      be somewhat challenging without increasing import requirements. -/
-    -- Porting note: rewrote proof to avoid removed constructor rat.mk_pnat
-    suffices
-      (Finset.range (x.num.natAbs * x.den)).sum
-          (Function.const ℕ ((1 : ℚ) / x.den * ((1 : ℚ) / x.den))) =
-        x
-      by exact this ▸ sum_mem fun n _ => AddSubmonoid.subset_closure ⟨_, rfl⟩
-    simp only [Function.const_apply, Finset.sum_const, Finset.card_range, nsmul_eq_mul]
-    rw [← Int.cast_ofNat, Int.ofNat_mul, Int.coe_natAbs,
-      abs_of_nonneg (num_nonneg_iff_zero_le.mpr hx), Int.cast_mul, Int.cast_ofNat]
-    rw [← mul_assoc, mul_assoc (x.num : ℚ), mul_one_div_cancel (Nat.cast_ne_zero.mpr x.pos.ne'),
-      mul_one, mul_one_div, Rat.num_div_den]
+@[simp] lemma addSubmonoid_closure_range_pow {n : ℕ} (hn₀ : n ≠ 0) (hn : Even n) :
+    closure (range fun x : ℚ ↦ x ^ n) = nonneg _ := by
+  refine le_antisymm (closure_le.2 <| range_subset_iff.2 hn.pow_nonneg) fun x hx ↦ ?_
+  suffices x = (x.num.natAbs * x.den ^ (n - 1)) • (x.den : ℚ)⁻¹ ^ n by
+    rw [this]; exact nsmul_mem (subset_closure <| mem_range_self _) _
+  rw [nsmul_eq_mul]
+  push_cast
+  rw [mul_assoc, pow_sub₀, pow_one, mul_right_comm, ← mul_pow, mul_inv_cancel, one_pow, one_mul,
+    ← Int.cast_ofNat, Int.coe_natAbs, abs_of_nonneg, ← div_eq_mul_inv, num_div_den]
+  rw [mem_nonneg] at hx
+  all_goals simp [x.den_pos.ne', Nat.one_le_iff_ne_zero, *]
```

Comment:
Can you prove this in terms of `NNRat.addSubmonoid_closure_range_pow` or vice, by composing both sides with a map / comap by the coercion?

Reply:
I don't see how because I can prove that `(closure (range fun x : α ↦ x ^ n)).map f ≤ closure (range fun x : β ↦ x ^ n)` if `f : α → β`, but that only tells me that `closure (range fun x : ℚ≥0 ↦ x ^ n) ≤ ⊤`.

Reply:
I quickly hacked this together. As is it's not currently better than what's written, but perhaps it could be cleaned up into something nice.

```lean
import Mathlib

open AddSubmonoid Set Rat NNRat

@[simp] lemma addSubmonoid_closure_range_pow {n : ℕ} (hn₀ : n ≠ 0) (hn : Even n) :
    AddSubmonoid.closure (range fun x : ℚ ↦ x ^ n) = nonneg _ := by
  have h₁ := NNRat.addSubmonoid_closure_range_pow hn₀
  let coeRat : ℚ≥0 →+ ℚ := { toFun := (↑), map_zero' := rfl, map_add' := by simp }
  have h₂ := AddMonoidHom.map_mclosure coeRat (range fun x ↦ x ^ n) |>.symm
  rw [h₁] at h₂
  convert h₂
  · ext x
    simp only [mem_range, AddMonoidHom.coe_mk, ZeroHom.coe_mk, mem_image, exists_exists_eq_and,
      coe_pow]
    constructor
    · rintro ⟨y, rfl⟩
      obtain ⟨k, rfl⟩ := (even_iff_exists_two_mul n).mp hn
      simp_rw [pow_mul]
      use ⟨|y|, by positivity⟩
      simp
    · rintro ⟨y, rfl⟩
      exact ⟨y, rfl⟩
  · ext x
    simp only [mem_nonneg, mem_map, mem_top, AddMonoidHom.coe_mk, ZeroHom.coe_mk, true_and]
    refine ⟨(⟨⟨x, ·⟩, rfl⟩), by rintro ⟨x, rfl⟩; exact coe_nonneg x⟩
```

Reply:
I managed to golf your proof dow to five lines. I think that's good enough!